paperqa/config.py

```diff
@@ -0,0 +1,413 @@
+    summary: str = summary_prompt
+    qa: str = qa_prompt
+    select: str = select_paper_prompt
+    pre: str | None = Field(
+        default=None,
+        description=(
+            "Opt-in pre-prompt (templated with just the original question) to append"
+            " information before a qa prompt. For example:"
+            " 'Summarize all scientific terms in the following question:\n{question}'."
+            " This pre-prompt can enable injection of question-specific guidance later"
+            " used by the qa prompt, without changing the qa prompt's template."
+        ),
+    )
+    post: str | None = None
+    system: str = default_system_prompt
+    use_json: bool = False
+    # Not thrilled about this model,
+    # but need to split out the system/summary
+    # to get JSON
+    summary_json: str = summary_json_prompt
+    summary_json_system: str = summary_json_system_prompt
+    model_config = ConfigDict(extra="forbid")
+    EXAMPLE_CITATION: ClassVar[str] = "(Example2012Example pages 3-4)"
+
+    @field_validator("summary")
+    @classmethod
+    def check_summary(cls, v: str) -> str:
+        if not set(get_formatted_variables(v)).issubset(
+            set(get_formatted_variables(summary_prompt))
+        ):
+            raise ValueError(
+                f"Summary prompt can only have variables: {get_formatted_variables(summary_prompt)}"
+            )
+        return v
+
+    @field_validator("qa")
+    @classmethod
+    def check_qa(cls, v: str) -> str:
+        if not set(get_formatted_variables(v)).issubset(
+            set(get_formatted_variables(qa_prompt))
+        ):
+            raise ValueError(
+                f"QA prompt can only have variables: {get_formatted_variables(qa_prompt)}"
+            )
+        return v
+
+    @field_validator("select")
+    @classmethod
+    def check_select(cls, v: str) -> str:
+        if not set(get_formatted_variables(v)).issubset(
+            set(get_formatted_variables(select_paper_prompt))
+        ):
+            raise ValueError(
+                f"Select prompt can only have variables: {get_formatted_variables(select_paper_prompt)}"
+            )
+        return v
+
+    @field_validator("pre")
+    @classmethod
+    def check_pre(cls, v: str | None) -> str | None:
+        if v is not None and set(get_formatted_variables(v)) != {"question"}:
+            raise ValueError("Pre prompt must have input variables: question")
+        return v
+
+    @field_validator("post")
+    @classmethod
+    def check_post(cls, v: str | None) -> str | None:
+        if v is not None:
+            # kind of a hack to get list of attributes in answer
+            attrs = set(Answer.model_fields.keys())
+            if not set(get_formatted_variables(v)).issubset(attrs):
+                raise ValueError(f"Post prompt must have input variables: {attrs}")
+        return v
+
+
+class AgentSettings(BaseModel):
+    agent_llm: str = Field(
+        default="gpt-4o-2024-08-06",
```

Comment:
It would be nice to move this repeated variable to a module-level default with a good name:

```python
# Set this to a model that has fast inference, low cost
# and decent performance, for use as a common default
FAST_CHEAP_DECENT_LLM = "gpt-4o-2024-08-06"
```

Reply:
Would rather leave in place - there may be another larger model soon that is too expensive to be in all three places